specs/LaunchingRegisteredProtocols.md

```diff
@@ -0,0 +1,318 @@
+  /// or replace the default dialog with a custom dialog.
+  ///
+  /// If a deferral is not taken on the event args, the external protocol launch is 
+  /// blocked until the event handler returns.  If a deferral is taken, the
+  /// external protocol launch is blocked until the deferral is completed.
+  /// The host also has the option to cancel the protocol launch.
+  ///
+  /// The `NavigationStarting` and `NavigationCompleted` events will be raised,
+  /// regardless of whether the `Cancel` property is set to `TRUE` or
+  /// `FALSE`. The `SourceChanged`, `ContentLoading`, and `HistoryChanged` events 
+  /// will not be raised regardless of this property. 
+  /// The `LaunchingExternalProtocol` event will be raised after the 
+  /// `NavigationStarting` event and before the `NavigationCompleted` event.
+  /// The default settings will also be updated upon navigation to an external
+  /// protocol. 
+  ///
+  /// If the request is made from a trustworthy origin
+  /// (https://w3c.github.io/webappsec-secure-contexts#potentially-trustworthy-origin) 
+  /// a checkmark box will be displayed on the default browser UI that gives the user 
+  /// the option to always allow the external protocol to launch from this origin.
+  /// If the user checks this box, upon the next request from that origin to the 
+  /// protocol, the event will still be raised, but there will be no default dialog shown
+  /// when `Cancel` is `FALSE`.  
+  ///
+  /// If the request is initiated by a cross-origin iframe without a user gesture, 
+  /// the request will be blocked and the `LaunchingExternalProtocol` event will not 
+  /// be raised.
+  /// \snippet SettingsComponent.cpp LaunchingExternalProtocol
+  HRESULT add_LaunchingExternalProtocol(
+      [in] ICoreWebView2LaunchingExternalProtocolEventHandler* eventHandler,
+      [out] EventRegistrationToken* token);
+
+  /// Remove an event handler previously added with
+  /// `add_LaunchingExternalProtocol`.
+  HRESULT remove_LaunchingExternalProtocol(
+      [in] EventRegistrationToken token);
+  }
+
+/// Receives the `LaunchingExternalProtocol` event.
+[uuid(e5fea648-79c9-47aa-8314-f471fe627649), object, pointer_default(unique)] 
+interface ICoreWebView2LaunchingExternalProtocolEventHandler: IUnknown {
+  /// Provides the event args for the corresponding event.
+  HRESULT Invoke(
+      [in] ICoreWebView2* sender,
+      [in] ICoreWebView2LaunchingExternalProtocolEventArgs* args);
+}
+
+/// Event args for `LaunchingExternalProtocol` event.
+[uuid(fc43b557-9713-4a67-af8d-a76ef3a206e8), object, pointer_default(unique)] 
+interface ICoreWebView2LaunchingExternalProtocolEventArgs: IUnknown {
+  /// The URI with the external protocol to be launched.
+
+  [propget] HRESULT Uri([out, retval] LPWSTR* value);
+
+  /// The origin initiating the external protocol launch.
+  /// The origin will be empty if the WebView2 navigated to the external protocol.

```

Comment:
I think this is saying that if the CoreWebView2.Navigate (or similar) method is used, or script in the web page changes the window.location, then there is no initiating origin.

If the initiation was due to the user clicking on a link, then the document that contains the clicked-on link is the initiating origin.

Did I get that right?

Reply:
Yes, there will also be an initiating origin if the request is initiated from an iframe.

Reply:
What about the case where a script in the top-level document assigns to `window.location`? Will that report the top-level document's Source as the InitiatingOrigin?

Reply:
Please make this text more explicit.

WebView2 navigated -> WebView2 navigated via CoreWebView2 APIs like ...

Also please explicitly state what happens if script causes the navigation with examples like eg `window.location = `